### PR TITLE
Fix bugs in backward. More/faster tests

### PR DESCRIPTION
**Pull request by danthe3rd**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #443
* #442
* #440
* #439
* #438
* #434
* #429
* #426
* #421
* #415
* #412
* #411
* #433
* #432
* #405
* #404
* #403
* #399
* #398
* #397
* #396
* __->__ #395

**TEST PLAN**
```
CUDA_VISIBLE_DEVICES=1 CUDA_LAUNCH_BLOCKING=1 time python -m pytest tests/test_mem_eff_attention.py
```

The tests run in ~22s on my machine now

### CODE REVIEW

File: tests/test_mem_eff_attention.py [hunk 1]
```diff
@@ -158,9 +156,15 @@ def test_logsumexp(
     batch_size,
     k_len,
     dtype,
-    op_kw: Tuple[xformers.ops.MemoryEfficientAttentionOp, Dict[str, Any]],
+    op: xformers.ops.MemoryEfficientAttentionOp,
 ):
-    op, kw = op_kw
+    if op.FORWARD_OPERATOR is None:
+        return
```
Comment (danthe3rd) [inline tests/test_mem_eff_attention.py:162]: (note: this skips flashattention)

File: xformers/components/attention/csrc/cuda/mem_eff_attention/kernel_forward.h [hunk 2]
```diff
@@ -652,7 +652,7 @@ struct AttentionKernel {
       if (query_start() + thread_id() < p.num_queries) {
         p.logsumexp_ptr[query_start() + thread_id()] =
             accum_t(mi[thread_id()]) + std::log(accum_t(s_prime[thread_id()]));
-      } else if (thread_id() < lse_dim) {
+      } else if (query_start() + thread_id() < lse_dim) {
         p.logsumexp_ptr[query_start() + thread_id()] =
```
Comment (danthe3rd) [inline xformers/components/attention/csrc/cuda/mem_eff_attention/kernel_forward.h:656]: Fixes a bug in logsumexp padding with batchSize > 1

File: xformers/components/attention/csrc/cuda/mem_eff_attention/kernel_backward.h [hunk 3]
```diff
@@ -503,17 +503,6 @@ struct AttentionBackwardKernel {
...
-      // Initialize shared-memory. It can contain `nans` otherwise that screw up
-      // everything (only seens on Sm75+ tho)
-      uint32_t* smem = (uint32_t*)smem_buffer;
-      for (int i = 0; i < sizeof(SharedStorage) / sizeof(uint32_t) -
-               kWarpSize * kNumWarpsPerBlock;
-           i += kWarpSize * kNumWarpsPerBlock) {
-        smem[i + thread_id] = 0;
-      }
-    };
-
```
Comment (danthe3rd) [inline xformers/components/attention/csrc/cuda/mem_eff_attention/kernel_backward.h:516]: no longer needed as we changed how we iterate on residuals - at least the tests pass without

File: tests/test_mem_eff_attention.py [hunk 4]
```diff
@@ -181,25 +187,12 @@ def test_logsumexp(
     assert_allclose(lse[:, : ref_lse.shape[1]], ref_lse, atol=2e-4)
 
 
-@pytest.mark.parametrize(
-    "attn_bias_type", [None, xformers.ops.LowerTriangularMask, torch.Tensor]
-)
+@pytest.mark.parametrize("attn_bias_type", [None, xformers.ops.LowerTriangularMask])
```
Comment (fmassa) [inline tests/test_mem_eff_attention.py:190]: nit: did you remove the test for `torch.Tensor` here?

File: xformers/ops.py [hunk 5]
```diff
@@ -86,17 +86,44 @@ class AttentionOpBase(torch.autograd.Function):
...
     SUPPORTED_MAX_K: float
     SUPPORTED_ATTN_BIAS_TYPES: Set[Any] = {type(None)}
     SUPPORTS_DROPOUT: bool
     NAME: str
 
+    _TEST_BATCH_SIZES: List[int] = [1, 300]
+    _TEST_K: List[int] = [32, 128]
+
+    @classmethod
+    def generate_test_shapes_B_Mq_Mkv_K(cls):
```
Comment (fmassa) [inline xformers/ops.py:109]: nit: I find it a bit weird that each operator has a method to generate test shapes. Not a blocker, but might be good to look into separating this afterwards, as we might also care about specific shapes for benchmarking

**Review by fmassa — APPROVED** [review]
Thanks!

I think there is a test case which is not being tested anymore, but apart from that LGTM!

**Pull request merged**